src/org/skypro/skyshop/basket/ProductBasket.java

```diff
@@ -3,57 +3,54 @@
 import org.skypro.skyshop.product.Product;
 
 import java.util.ArrayList;
-import java.util.Iterator;
+import java.util.HashMap;
+import java.util.Map;
 
 public class ProductBasket {
 
-    private final ArrayList<Product> products;
+    private final Map<String, ArrayList<Product>> products;
 
     public ProductBasket() {
-        products = new ArrayList<>();
+        products = new HashMap<>();
     }
 
     public void add(Product product) {
-        products.add(product);
+        String key = product.getName();
+        ArrayList<Product> productListByKey;
+        if (products.containsKey(key)) {
+            productListByKey = products.get(key);
+        } else {
+            productListByKey = new ArrayList<>();
+        }
+        productListByKey.add(product);
+        products.put(key, productListByKey);
         System.out.println("Продукт добавлен в корзину.");
     }
 
-    public ArrayList<Product> del(String nameOfProduct){
-        ArrayList<Product> removedProducts = new ArrayList<>();
-        Iterator<Product> productIterator = products.iterator();
-        while (productIterator.hasNext()) {
-            Product nextproduct = productIterator.next();
-            if (nextproduct.getName().equals(nameOfProduct)) {
-                removedProducts.add(nextproduct);
-                productIterator.remove();
-            }
-        }
+    public ArrayList<Product> del(String nameOfProduct) {
```

Comment:
Соблюдаем абстракцию
```suggestion
    public List<Product> del(String nameOfProduct) {
```